File: src/main/clients/openrgb/effects/StarryNight.ts
import { RGBColor } from '@main/clients/openrgb/client/classes/RGBColor';
import { AbstractEffect } from '@main/clients/openrgb/effects/base/AbstractEffect';

class StarryNight extends AbstractEffect {
  public constructor() {
    super('Starry Night', false);
  }

  private getRandom(): RGBColor {
    return RGBColor.fromHSV(Math.floor(Math.random() * 360), 1, 1);
  }

  private maxSteps = 30;

  protected async applyEffect(): Promise<void> {
    const promises: Array<Promise<void>> = [];
    for (let i = 0; i < this.devices.length; i++) {
      promises.push(
        new Promise<void>((resolve) => {
          (async (): Promise<void> => {
            const leds: Array<RGBColor> = Array(this.devices[i].leds.length).fill(
              RGBColor.fromHex('#000000')
            );
            const steps: Array<number> = Array(this.devices[i].leds.length).fill(0);
            for (let iter = 0; this.isRunning; iter = (iter + 1) % this.maxSteps) {
              const newColors = Array(this.devices[i].leds.length);

              for (let j = 0; j < leds.length; j++) {
                steps[j] = Math.max(0, steps[j] - 1);
                newColors[j] = leds[j].getDimmed(steps[j] / this.maxSteps);
              }

              const canTurnOn = steps.filter((i) => i > 0).length / steps.length < 0.1;
              if (iter == 0 || canTurnOn) {
                let ledOn = -1;
                do {
                  ledOn = Math.floor(Math.random() * leds.length);
                } while (steps[ledOn] > 0);
                steps[ledOn] = 15 + Math.floor(Math.random() * 15);
                newColors[ledOn] = leds[ledOn] = this.getRandom().getDimmed(
                  steps[ledOn] / this.maxSteps
                );
              }

              this.setColors(this.devices[i], newColors);
              await this.sleep(Math.random() * 150);
            }
            resolve();
          })();
        })
      );
    }

    await Promise.all(promises);
    this.hasFinished = true;
  }
}

export const starryNight = new StarryNight();

File: lib/clients/openrgb/effects/starry_night.py
from .base.abstract_effect import AbstractEffect
from ..sdk.utils import RGBColor
from ....utils.singleton import singleton

import random
import threading


@singleton
class StarryNight(AbstractEffect):
    def __init__(self):
        super().__init__("Starry night")

    def get_random(self):
        hue = random.randint(0, 359)
        return RGBColor.fromHSV(hue, 1, 2)

    def apply_effect(self):
        self.max_steps = 30
        self.threads = []

        for device in self.devices:
            thread = threading.Thread(
                name=f"StarryNight-dev-{device.id}",
                target=self.device_effect,
                args=(device,),
            )
            self.threads.append(thread)
            thread.start()

        for thread in self.threads:
            thread.join()

    def device_effect(self, device):
        leds = [RGBColor.fromHEX("#000000")] * len(device.leds)
        steps = [0] * len(device.leds)

        while self.is_running:
            new_colors = [None] * len(device.leds)

            for i in range(len(leds)):
                steps[i] = max(0, steps[i] - 1)
                new_colors[i] = leds[i].dim(steps[i] / self.max_steps)

            can_turn_on = sum(1 for i in steps if i > 0) / len(steps) < 0.1

            if can_turn_on:
                led_on = -1
                while led_on < 0 or steps[led_on] > 0:
                    led_on = random.randint(0, len(leds) - 1)
                steps[led_on] = 15 + random.randint(0, 15)
                new_colors[led_on] = leds[led_on] = self.get_random().dim(
                    steps[led_on] / self.max_steps
                )

            self.set_colors(device, new_colors)
            self.sleep(random.randint(0, 150) / 1000)


starry_night = StarryNight()
